tests/lib/in_memory/database.py

```diff
@@ -344,3 +302,58 @@ def parse_value(value):
 def nulls_first_order(key):
     # Usable as a key function to `sorted()` which sorts NULLs first
     return (0 if key is None else 1, key)
+
+
+def apply_function_to_columns(fn, *columns):
+    reshaped_columns = reshape_columns(*columns)
+    patients = set().union(*[c.patients() for c in reshaped_columns])
+
+    return Column(
+        {
+            p: [
+                fn(*args)
+                for args in zip(*[column.get_values(p) for column in reshaped_columns])
+            ]
+            for p in patients
+        },
+        default=fn(*[column.default for column in reshaped_columns]),
+    )
+
+
+def reshape_columns(column, *other_columns):
+    reshaped_others = []
+    for other in other_columns:
+        column, other = reshape_column_pair(column, other)
+        reshaped_others.append(other)
+    return [column] + reshaped_others
```

Comment:
Gah! You're quite right. I don't think this will work. I naively hoped I could convert the existing pairwise code into something more general without having to actually understand what it was doing. I think it's back to the drawing board with this.

Reply:
I _think_ I've got this right now